flax/linen/module.py

```diff
@@ -790,9 +791,52 @@ def make_rng(self, name: str) -> PRNGKey:
       raise ValueError("Can't use RNGs on unbound modules")
     return self.scope.make_rng(name)
 
+  def bind(self, variables: VariableDict, *args, rngs: RNGSequences = None,
+           mutable: CollectionFilter = False):
+    """Creates an interactive Module instance by binding variables and RNGs.
+
+    bind provides an "interactive" instance of a Module directly without
+    transforming a function with ``apply``. This is particulary useful for debugging
+    and interactive use cases like notebooks where a function would limit the ability
+    split up code into different cells.
+
+    Once the variables (and optionally RNGs) are bound to a ``Module`` it becomes a
+    stateful object. Note that idiomatic JAX is functional and therefore an interactive
```

Comment:
I think this statement isn't true if `mutable=False` -- you're not able to shoot yourself in the foot (correct me if I am wrong)

If you agree, would be good to clarify in the comment.

Reply:
In practise, I think it doesn't matter that much. Normally you can train a model or compile a train step with `mutable=False` but you cannot combine this api with optimizers or `jax.jit`. The pitfall I describe here is about composability with Jax and other apis that expect functional objects which still holds for `mutable=False`.

Reply:
Isn't this fine?
```
@jax.jit
def foo(params, inputs):
  return CNN().bind({"params": variables})(inputs)
  
jax.grad(foo)
```
And then `foo` works fine with `jax.grad` etc?

Reply:
Yes but here you are basically inlining apply. The bound Module lives inside a function. The problem is that you can keep a bound module around and just pass the bound module around until you start using jax transforms.

Reply:
So of course we could tell users to bind only inside functions but it seems better to me to stick to a single pattern `apply` to avoid having multiple ways to do the same thing.

Reply:
Ah. I guess there are two different things:

1. Binding a module within a function. This works fine also with mutable, e.g.:

```
@jax.jit
def prediction(params, batch_stats, inputs):
  cnn = CNNWithBatchNorm().bind({"params": variables, "batch_stats", batch_stats}, mutable=['batch_stats'])
  outputs = cnn(inputs)
  return outputs, cnn.variables['batch_stats']
```
  
2. Colab style tinkering, e.g.:
 
```
cnn = CNNWithBatchNorm().bind(...)
  
@jax.jit
def predict(inputs):
  return cnn(inputs)
```

IIUC this would technically work even if you use `mutate=False`, but it would compile the variables in `cnn` as constants when jitting `predict` which would make the compiled HLO much larger, and would be unsafe if anyone modifies `cnn` or `cnn.variables` later on. So maybe this is what your comment was referring to.

Reply:
I guess the real thing is that to use JAX safely/effectively you never want to:
1. Leak a bound module "out" of a function, e.g. in example 1 above, if you would return `cnn` from the `prediction` function; or
2. Leak a bound module "in" to a function (e.g. closing over a bound instance e.g. example 2 above)

Reply:
Thanks, I understand now. I'll resolve the conversation.

Reply:
Would it be possible/desirable to avoid above pitfalls by throwing an error when a `module.bind()` is used inside a JAX transformation and/or refusing to be called from a Python module (as opposed to an interactive environment)?